### PR TITLE
parachain/disputes: import module for disputes-coordinator

### PR DESCRIPTION
**Pull request by kanishkatn**
## Changes

Implements the imports module for disputes-coordinator. Please find the polkadot implementation here: https://github.com/paritytech/polkadot/blob/master/node/core/dispute-coordinator/src/import.rs

```
Vote import logic.

This module encapsulates the actual logic for importing new votes and provides easy access of
the current state for votes for a particular candidate.

In particular there is `CandidateVoteState` which tells what can be concluded for a particular
set of votes. E.g. whether a dispute is ongoing, whether it is confirmed, concluded, ..

Then there is `ImportResult` which reveals information about what changed once additional votes
got imported on top of an existing `CandidateVoteState` and reveals "dynamic" information, like
whether due to the import a dispute was raised/got confirmed, ...
```

There are no tests for this module in polkadot but it is part of the coordinator tests. I'll be handling it there itself and also see if we can write tests for this module later. 

This can be reviewed again once it is integrated into the coordinator. I'd appreciate an early review but feel free to skip it.

## Issues
#3302

## Primary Reviewer

@EclesioMeloJunior

### CODE REVIEW

**Review by kishansagathiya — COMMENTED** [review]
I struggle to review PRs like this.

- This adds import.go, but it is not getting used anywhere. If we could see where it used, we could see what it is used for and if it implements the functionality that it is expected to. Right now, only thing I can do is to see if it translates rust's import.rs file into go.
- We should be looking at rust code to understand what functionalities we have to implement and not translate individual files directly. This way we might end up adding things that we might not need later.


I would suggest creating a skeleton PR for dispute coordinator against `feat/parachain`. 
- Create a struct for the subsystem say, `DisputeCoordinator`, with channel to and from Overseer in it
- Implement SubSystem interface for it. Implement `Run` method ( You can see the functionality for that here https://github.com/paritytech/polkadot-sdk/blob/0a6dfdf973b8e7b669eda6e2ed202fb3549a20b9/polkadot/node/core/dispute-coordinator/src/lib.rs#L141)
- Inside `Run` method process messages received from overseer (overseer signal and subsystem messages) say in `processMsg` https://github.com/paritytech/polkadot-sdk/blob/0a6dfdf973b8e7b669eda6e2ed202fb3549a20b9/polkadot/node/core/dispute-coordinator/src/initialized.rs#L253-L254 
- For message processing, create all the necessary message types (Initially just empty structs and add fields as needed). These struct don't need to be same as in rust as they are only for intra-node communication.
- Inside Run method, process messages from participation receiver. Don't have to implement it's full functionality. Create message types and functions with TODOs in it. Those TODO can be done later as individual issues.

You could look at https://github.com/ChainSafe/gossamer/pull/3497 as an example of a skeleton PR. 

With this approach, I will have a lot more clarity about the code and work. As an added benefit, it would be easier to review, at least for me.

I can create an issue for this if there is alignment.
Tagging @timwu20 and @EclesioMeloJunior for thoughts!

**Comment by kanishkatn** [comment]
> I struggle to review PRs like this.
> 
> * This adds import.go, but it is not getting used anywhere. If we could see where it used, we could see what it is used for and if it implements the functionality that it is expected to. Right now, only thing I can do is to see if it translates rust's import.rs file into go.
> * We should be looking at rust code to understand what functionalities we have to implement and not translate individual files directly. This way we might end up adding things that we might not need later.
> 
> I would suggest creating a skeleton PR for dispute coordinator against `feat/parachain`.
> 
> * Create a struct for the subsystem say, `DisputeCoordinator`, with channel to and from Overseer in it
> * Implement SubSystem interface for it. Implement `Run` method ( You can see the functionality for that here https://github.com/paritytech/polkadot-sdk/blob/0a6dfdf973b8e7b669eda6e2ed202fb3549a20b9/polkadot/node/core/dispute-coordinator/src/lib.rs#L141)
> * Inside `Run` method process messages received from overseer (overseer signal and subsystem messages) say in `processMsg` https://github.com/paritytech/polkadot-sdk/blob/0a6dfdf973b8e7b669eda6e2ed202fb3549a20b9/polkadot/node/core/dispute-coordinator/src/initialized.rs#L253-L254
> * For message processing, create all the necessary message types (Initially just empty structs and add fields as needed). These struct don't need to be same as in rust as they are only for intra-node communication.
> * Inside Run method, process messages from participation receiver. Don't have to implement it's full functionality. Create message types and functions with TODOs in it. Those TODO can be done later as individual issues.
> 
> You could look at #3497 as an example of a skeleton PR.
> 
> With this approach, I will have a lot more clarity about the code and work. As an added benefit, it would be easier to review, at least for me.
> 
> I can create an issue for this if there is alignment. Tagging @timwu20 and @EclesioMeloJunior for thoughts!

Thank you for your suggestion @kishansagathiya .

I believe we already spoke about this and I am sorry you're struggling to review this PR. The coordinator is almost complete; pending tests. Feel free to skip this PR.

**Comment by kanishkatn** [comment]
@kishansagathiya Also, like I mentioned before, I'm happy to get on a call to take you again through the disputes-coordinator module, the approach I'm taking and sort out any questions/concerns you have!

**Comment by kanishkatn** [comment]
> I agree it doesn't make much sense to to review code that isn't referenced at this point. I understand the approach of translating individual files, but without tests, and or any references to this code it makes it difficult for the reviewer to understand the context. We will need to do a more exhaustive code review on the feature branch `feat/k/disputes-coordinator` after this PR gets merged into it.

Thank you for sharing your thoughts @timwu20 . I am going to keep this PR open to get some early feedback from @EclesioMeloJunior for this `module` if he has the time. The rest of the team can feel free to skip it.

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/parachain/dispute/types/vote.go [hunk 1]
```diff
@@ -268,11 +276,40 @@ func NewCandidateVoteStateFromReceipt(receipt parachainTypes.CandidateReceipt) (
...
+type ValidCandidateVotes map[parachainTypes.ValidatorIndex]Vote
+
+func (vcv ValidCandidateVotes) InsertVote(vote Vote) (bool, error) {
+	_, ok := vcv[vote.ValidatorIndex]
+	if !ok {
+		vcv[vote.ValidatorIndex] = vote
+		return true, nil
+	}
+
+	existingVote := vcv[vote.ValidatorIndex]
```
[the remark this answers is not in the document]
Comment (kanishkatn) [reply dot/parachain/dispute/types/vote.go:289]: I don't know what I was thinking, haha. Thank you!

**Review by EclesioMeloJunior — APPROVED** [review]
LGTM

**Pull request closed without merging**